### PR TITLE
[WIP] Security analysis : get logs even when an exception is raised

### PR DESCRIPTION
**Pull request by yichen88**
In the scenario with log, the result of computation may be null.

Signed-off-by: yichen88 <tang.yichenyves@gmail.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: security-analysis/security-analysis-api/src/main/java/com/powsybl/security/distributed/DistributedSecurityAnalysis.java [hunk 1]
```diff
@@ -149,10 +150,19 @@ private String getLogOutputName(int taskNumber) {
 
         @Override
         public SecurityAnalysisResultWithLog after(Path workingDir, ExecutionReport report) throws IOException {
-            SecurityAnalysisResult re = simpleSubTaskHandler.after(workingDir, report);
+            SecurityAnalysisResult re = null;
+            boolean error = false;
+            try {
+                re = simpleSubTaskHandler.after(workingDir, report);
+            } catch (PowsyblException e) {
+                error = true;
```
Comment (mathbagu) [inline security-analysis/security-analysis-api/src/main/java/com/powsybl/security/distributed/DistributedSecurityAnalysis.java]: Should we log this error?

**Comment by sylvlecl** [comment]
@yichen88 :
After thinking again, I am wondering how we should handle those errors. I think we must clarify the SecurityAnalysis interface contract.
With the proposed evolution, when an error occurs in 1 subcommand of distributed security analysis :
* the future returned by `run` will throw an execution exception
* the future returned by `runWithLog` will not throw, but will returns an empty result with (maybe) logs. The info from `ExecutionReport` is kind of lost here for the caller (which executions worked well, which did not).

Also I am not sure the evolution is enough, because we don't handle null results in the `SecurityAnalysisTool`. What should we do in that case : should we write down the logs and return an exit code != 0 ?

What would you think about returning a more complete object for the `runWithLog` call:
 - keeping a reference to the exception in case of exception
 - maybe throwing it when calling "getResult"
 - having an execution status "failed/succeeded"
 - still having a reference to the logs
We may call it SecurityAnalysisExecutionResult, which is less specific.

Another approach would be to keep the call as is, but to throw a specific type of exception embedding the source exception together with a reference to the logs. Then the client code would be in charge of handling that exception to retrieve the logs.

**Comment by geofjamg** [comment]
> 
> 
> @yichen88 :
> After thinking again, I am wondering how we should handle those errors. I think we must clarify the SecurityAnalysis interface contract.
> With the proposed evolution, when an error occurs in 1 subcommand of distributed security analysis :
> 
>     * the future returned by `run` will throw an execution exception
> 
>     * the future returned by `runWithLog` will not throw, but will returns an empty result with (maybe) logs. The info from `ExecutionReport` is kind of lost here for the caller (which executions worked well, which did not).
> 
> 
> Also I am not sure the evolution is enough, because we don't handle null results in the `SecurityAnalysisTool`. What should we do in that case : should we write down the logs and return an exit code != 0 ?
> 
> What would you think about returning a more complete object for the `runWithLog` call:
> 
>     * keeping a reference to the exception in case of exception
> 
>     * maybe throwing it when calling "getResult"
> 
>     * having an execution status "failed/succeeded"
> 
>     * still having a reference to the logs
>       We may call it SecurityAnalysisExecutionResult, which is less specific.
> 
> 
> Another approach would be to keep the call as is, but to throw a specific type of exception embedding the source exception together with a reference to the logs. Then the client code would be in charge of handling that exception to retrieve the logs.

For me, there is a serious design issue with this. It is the purpose of Completable future to embed exception handling and I think it would be a mistake to store exception somewhere in the security analysis result data structure.
Also about the PR, I would not allow null result. Again it is the CompletableFuture job to carry result or not depending on execution.

**Comment by sylvlecl** [comment]
> For me, there is a serious design issue with this. It is the purpose of Completable future to embed exception handling and I think it would be a mistake to store exception somewhere in the security analysis result data structure.
> Also about the PR, I would not allow null result. Again it is the CompletableFuture job to carry result or not depending on execution.

@geofjamg : just seeing your feedback, thank you. However, what solution would you propose ? Catch the exception and only log it, which makes it not accessible to the user? Or include the logs inside the exception, then inspect it to retrieve the logs? 
I changed the title of the PR to make the goal clear.

As a general design interrogation, I wonder if we should really try to stick so hard to the basic `CompletableFuture` API for our computations. Why not having objects more dedicated to our computation needs? (remember questions already raised about getting the status of computations etc).

@yichen88 :  whatever we choose, it will need to be "refactored" after PR #778 .

**Comment by sylvlecl** [comment]
Summary of discussions between @geofjamg @yichen88 and @sylvlecl :

The preferred solution would finally be, in case of exception, to have the logs carried by the exception (create a dedicated exception type).
Pros compared to alternative solutions :
 * the `run` and `runWithLog` methods stay consistent, in that they either complete exceptionally or return a non-null security analysis result
 * the underlying exceptions may be handled by the user (for display or simply propagation for example)
 * the logs are accessible in both "success" and "failure" cases (but in 2 different ways)

But we must pay attention to :
 * describe well in javadoc and/or through a helper method how to get the logs in case of exception
 * should this exception be serializable ?

Alternative solutions were:
 * optional result in `SecurityAnalysisResultWithLog` but this breaks consistency between the 2 methods
 * non null guarantee on `SecurityAnalysisResult` with a new status saying it encountered a "technical error". In this case the 2 methods would never complete exceptionally, but this means the underlying exceptions are lost (just logged) and cannot be handle by the user (for display or whatever)

@mathbagu Could you please provide an opinion about this solution ?

**Comment by yichen88** [comment]
Refactor in #807

**Pull request closed without merging**